### DISCUSSION TITLE
Minion requires superuser on Postgres 15?

### DISCUSSION THREAD

**Discussion by wozman** [mojolicious/mojo#1995] [Q&A]
Hi,

I did a fresh install of Mojolicious and Minion under perl 5.36 earlier and also installed Postgres 15. I created a new database just for Minion, created a standard user, but when I try to start up a worker I'm getting the following error: -

 `DBD::Pg::st execute failed: ERROR:  relation "minion_workers" does not exist at /Users/xxx/Library/Perl/perls/perl-5.36.0/lib/site_perl/5.36.0/Minion/Backend/Pg.pm line 255.`

If I grant the superuser role to my minion user then everything works. Is this a Postgres 15 issue? I don't remember having to grant superuser before.

UPDATE: Not a Minion issue, things have changed in PG15 - see https://www.postgresql.org/docs/15/ddl-schemas.html#DDL-SCHEMAS-PATTERNS

Woz

**Comment by kraih** [comment]
Wonder if there's anything we can do to improve the situation.

**Reply by wozman** [reply]
Actually it's pretty simple to solve after all once you read up on Postgres schemas/search paths.

By default the schema search_path is "$user", public, so create commands etc in the migrations will attempt to create objects in the connected users schema, or if that does not exist, in the public schema.

So there are two solutions I've found so far.

1) Assuming you user is 'minion' the quick and easy method is to grant create access to minion in the public schema

```
GRANT CREATE ON SCHEMA public to minion;
```

2) Create a minion schema on your database and grant create access to the minion user to that minion schema

```
CREATE SCHEMA minion AUTHORIZATION minion;
GRANT CREATE ON SCHEMA minion TO minion;
```

There may well be better ways, but my database security fu is weak!

So it's really just a case of documenting how to setup your Postgres DB correctly to allow the migrations to work.

**Comment by s1037989** [comment]
I would love to see some Minion / Mojo::Pg documentation for a simple "get
postgres up and running from nothing" just to supplement the hello world
examples so that the examples and even a simple POC is easy and fast to do.

I love postgres, but when I install from apt, I always have a few basic
things I have to remember to do each time when I'm on a new system and that
just really slows me down and sometimes has me give up before I started. If
I'm really motivated I'll use Mojo::SQLite to avoid those barriers.

I would love to start my POCs with postgres more often, but I have an idea
that I want to code right now and I can't be bothered with postgres right
now, y'know?

**Reply by wozman** [reply]
Yes, definitely, and having just been through this I'll post this here for reference.

Create a file called, say, `create_minion_db.sql` with the following contents: -

```
CREATE DATABASE minion;
\connect minion;
CREATE USER minion WITH ENCRYPTED PASSWORD 's3cret';
GRANT ALL PRIVILEGES ON DATABASE minion TO minion;
CREATE SCHEMA minion AUTHORIZATION minion;
GRANT CREATE ON SCHEMA minion TO minion;
```

Then run that using: -

```
psql -f create_minion_db.sql postgres
```

Which will connect to Postgres as the default postgres user which should already exist after installing with apt and create a database called 'minion' with a 'minion' user and 'minion' schema with the correct permissions for the migrations to be able to create all the minion tables etc.

**Comment by s1037989** [comment]
Thanks for this! One follow up: I always seem to have to modify one or two
config files after installing. The key one is the ACL. When I'm in a hurry
and just working locally, I find the easiest thing to do is set the local
connection to "trust". I feel like I usually have to modify something in
postgres.conf as well as pg_hba.conf, or something like that. Clearly I
haven't done it too recently...
